Containerize

Nathan-Hausman opened the pull request:
The website is now containerized and works on the Pi.

Honestly there isn't too much to explain, the Dockerfile is really simple. If you want to test it out, run the following on the pi:
`sudo docker run -d -p 28070:28070 --name <name> website:test`

You should be able to see it run with `sudo docker ps`.

Nathan-Hausman commented:
Btw, I forgot to update the README.md, I will do that later today or tomorrow and push it as soon as this is approved.

markser commented:
LGTM

Some weird things with the pr, the pr seems to be missing `docker-entrypoint.sh` for some reason but `docker-entrypoint.sh` shows up in the git commit `daa115f`. Could be something on my end, no big deal I just checkout the latest commit. 

Some Nitpicks
- Lets say that I wanted to change the port number, what would need to be done to do that? Assuming that the other containers/pods are using that port number what would happen? I dont think we need something that dynamically assign the port values as it could take alot of time/resources to search through the ip tables. (Sidenote, we probably want something that checks the available ports so that we aren't messing other pods/containers that are deployed with an EXPOSED/PUBLISHED port if it's not done so already).
  - This looks promising: https://stackoverflow.com/a/28721643
  - https://www.cloudsavvyit.com/14880/whats-the-difference-between-exposing-and-publishing-a-docker-port/
- Can we add within the readme on how to build the image, would be nice to have just in case
- Some comments on what some of the lines on the Dockerfile might help if we have to troubleshoot down the line.

detjensrobert commented on Dockerfile:
```diff
@@ -0,0 +1,5 @@
+FROM arm64v8/node:17.3-bullseye
```
This should ideally use an architecture-agnostic node image, if such exists. 

Looks we can use https://docs.docker.com/buildx/working-with-buildx/ to make a multi-platform image, I can look into implementing that for our build process.

Nathan-Hausman replied:
I didn't use one because this will only be running on Pi devices for our pipeline. I can change it, but I didn't think it was necessary with our use case.

detjensrobert reviewed (approved):
Was there a reason to move the website into a subdir? It makes more sense to me to keep it in the root.

detjensrobert reviewed (changes requested):
that was supposed to be a request not approval oops

Nathan-Hausman commented:
> Some weird things with the pr, the pr seems to be missing `docker-entrypoint.sh` for some reason but `docker-entrypoint.sh` shows up in the git commit `daa115f`. Could be something on my end, no big deal I just checkout the latest commit.

Yeah I removed it, it exists in the earlier build stages of the node image, I briefly misunderstood and added my own script, then removed it later, it looks like I forgot to put that in the commit message, my bad.


> Lets say that I wanted to change the port number, what would need to be done to do that? Assuming that the other containers/pods are using that port number what would happen? I dont think we need something that dynamically assign the port values as it could take alot of time/resources to search through the ip tables. (Sidenote, we probably want something that checks the available ports so that we aren't messing other pods/containers that are deployed with an EXPOSED/PUBLISHED port if it's not done so already).

Yeah I'll look into this! 

__Edit:__ Sorry, a bit out of it today, if we want to have it be exposed on a different port, we can publish it in the `docker run` command to connect the container's port to a different real port. For example, `-p 28070:80`. I'm adding this into the readme right now. As for expose, I initially had it in there mainly for documentation purposes, but it was causing issues with the build (which was really weird), but let me try adding it again.

>     * Can we add within the readme on how to build the image, would be nice to have just in case
>     * Some comments on what some of the lines on the Dockerfile might help if we have to troubleshoot down the line.

Yeah totally can do, I'm sorry I am behind, last week was brutal on me. I'm gonna work extra hard this week to make up.

Nathan-Hausman commented:
> Was there a reason to move the website into a subdir? It makes more sense to me to keep it in the root.

It makes it easier to move together into the image? I mean I can do it either way, it was just easier this way. If there is a reason to change it back I don't mind, but I don't think it matters either way

Nathan-Hausman commented:
Okay, I have added some basic command line options to change the host and port of the Node server, and updated README.md and added comments to the Dockerfile.

markser commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+FROM arm64v8/node:17.3-bullseye
+EXPOSE tcp/28070
```
[NOT BLOCKING]

From what I've seen, it looks like portnumber/tcp is the common practice
It works, but I was wondering if there was any reason for tcp/28070? 

https://docs.docker.com/config/containers/container-networking/
https://www.ctl.io/developers/blog/post/docker-networking-rules

markser commented on website/server.js:
```diff
@@ -0,0 +1,32 @@
+var parseArgs = require('minimist');
+var argv = parseArgs(process.argv.slice(2), opts={
+    "string": "host",
+    "default": {"port": 28070}
```
This might limit our dynamic port selection dreams. Let's just drop that idea if you are still tinkering with it and just deal with it as it comes up as we have documentation on it now :)

markser commented:
LGTM
No worries about the port values if we are having troubles with it. 
What I'm going to assume will happen when this is deployed on our platform is that the other docker containers/kubernetes pods will run into a port error rather than the website. 
I am going to approve as we have abstracted the build platforms to our pipeline instead.
